### PR TITLE
Add bar background clipping

### PR DESCRIPTION
**Pull request by izeau**
Implements the features discussed in #274. It allows the clipping of the bar background to reveal what is underneath. The clipping shape is based on the background dimensions & corner radius, and the clipping mask is a float (`0.f` = no clipping, `1.f` = fully clipped). The `background.clip` property is fully animatable, and can be applied to items and brackets, but not to the bar itself or to text items.

Floating point number properties can now be animated using the `ANIMATE_FLOAT` macro.

Clipping should not affect performance when disabled, aside from an additional loop in the `bar_draw` function. When enabled, there are at the moment quite a few preventable bar redraws since the `bar_needs_refresh` is naive and does not check if the bar actually needs a refresh (i.e. the clipping value has changed, the item has moved or its dimensions were changed). A less aggressive change detection may be implemented but would require a bit of refactoring to the update functions.

### CODE REVIEW

**Comment by izeau** [comment]
Just noticed a bug: I'm not sure why but when using different display, the width of a bracket displaying the list of spaces is always the same, no matter the number of spaces.

**Comment by FelixKratz** [comment]
I believe we can get rid of some of the checks and unnecessary draw calls for the bar by changing some of the logic. We can do all the handlling of the clipping inside of the `bar_draw` and `{bar_item | backgound | group}_calculate_bounds` functions I believe, since we have all the information about wether an item has actually changed its background frame. If you give me push rights to the pull-request I can try changing it. As it is currently having a stationary clip creates drastic draw call overhead (from ~0.1% CPU to ~0.5% CPU with `draw_rect` taking the majority of CPU time).

**Comment by izeau** [comment]
Thanks for taking the time to look into it! You should have the rights to push updates to the branch already.

>We can do all the handlling of the clipping inside of the `bar_draw` and `{bar_item | backgound | group}_calculate_bounds` functions I believe, since we have all the information about wether an item has actually changed its background frame.

Do we have up-to-date info on `background.corner_radius` in these locations?

**Comment by FelixKratz** [comment]
I have pushed some changes, now clipping only happens exactly when needed and works for icon and label backgrounds as well. I have not done exhaustive testing on this, so before merging we should check that everything is working correctly.

**Comment by izeau** [comment]
This is great! Nothing broke in my config, and clipping may be animated still -- I tested varying the alpha value, `y_offset`, `label.width`, and `background.corner_radius`.

Thanks again

**Comment by FelixKratz** [comment]
Great, I have one more improvement in mind that will skip the loop if no item uses clip, but then it is ready I think.

**Review by FelixKratz — APPROVED** [review]

**Comment by FelixKratz** [comment]
Thank you for the PR!

**Pull request merged**
